src/Microsoft.ML.FastTree/GamModelParameters.cs

```diff
@@ -0,0 +1,911 @@
+                0, binIndices.Length - 1, internalNodeIndices, lteChild, gtChild, ref internalNodeId);
+            // internalNodeId should have been counted all the way down to 0 (root node)
+            Host.Assert(internalNodeId == 0);
+
+            var tree = (
+                thresholds: internalNodeIndices.Select(x => (float)binThresholds[binIndices[x]]).ToArray(),
+                lteChild: lteChild.ToArray(),
+                gtChild: gtChild.ToArray());
+            return tree;
+        }
+
+        private int CreateBalancedTreeRecursive(int lower, int upper,
+            List<int> internalNodeIndices, List<int> lteChild, List<int> gtChild, ref int internalNodeId)
+        {
+            if (lower > upper)
+            {
+                // Base case: we've reached a leaf node
+                Host.Assert(lower == upper + 1);
+                return ~lower;
+            }
+            else
+            {
+                // This is postorder traversal algorithm and populating the internalNodeIndices/lte/gt lists in reverse.
+                // Preorder is the only option, because we need the results of both left/right recursions for populating the lists.
+                // As a result, lists are populated in reverse, because the root node should be the first item on the lists.
+                // Binary search tree algorithm (recursive splitting to half) is used for creating balanced tree.
+                var mid = (lower + upper) / 2;
+                var left = CreateBalancedTreeRecursive(
+                    lower, mid - 1, internalNodeIndices, lteChild, gtChild, ref internalNodeId);
+                var right = CreateBalancedTreeRecursive(
+                    mid + 1, upper, internalNodeIndices, lteChild, gtChild, ref internalNodeId);
+                internalNodeIndices.Insert(0, mid);
+                lteChild.Insert(0, left);
+                gtChild.Insert(0, right);
+                return --internalNodeId;
+            }
+        }
+        private static RegressionTree CreateRegressionTree(
+            int numLeaves, int[] splitFeatures, float[] rawThresholds, int[] lteChild, int[] gtChild, double[] leafValues)
+        {
+            var numInternalNodes = numLeaves - 1;
+            return RegressionTree.Create(
+                numLeaves: numLeaves,
+                splitFeatures: splitFeatures,
+                rawThresholds: rawThresholds,
+                lteChild: lteChild,
+                gtChild: gtChild.ToArray(),
+                leafValues: leafValues,
+                // Ignored arguments
+                splitGain: new double[numInternalNodes],
+                defaultValueForMissing: new float[numInternalNodes],
+                categoricalSplitFeatures: new int[numInternalNodes][],
+                categoricalSplit: new bool[numInternalNodes]);
+        }
+
+        /// <summary>
+        /// The GAM model visualization command. Because the data access commands must access private members of
```

Comment:
>must access private members [](start = 82, length = 27)

I think it's a sign what we haven't properly design predictor.
From what I see it needs:
`_inputType.ValueCount` which is same as `_numFeatures` which I think we need to expose.
`_inputFeatureToDatasetFeatureMap` , `_binEffects` and `_featureMap` First two I think you use in `GetFeatureBinUpperBounds` and it can probably be rewritten. last one can be made as internal.

I found it's strange what we make command line thing which has more functionality than user can achieve by himself.

Reply:
Point of confusion. Currently, there can be fewer shape functions built than input features because of flocking (I think this is a bug and will file one separately.)

`_inputType.ValueCount` is _different_ than `_numFeatures`. I've rename `numFeatures` to `NumShapeFunctions` to clarify this.

`inputFeaturesToDatasetFeatureMap` has been renamed to `_inputFeatureToShapeFunctionMap`

`featureMap` has been renamed to `shapeToInputMap`. (and is `private`).

> I found it's strange what we make command line thing which has more functionality than user can achieve by himself.

 You mean the visualization command? With the new public interfaces, the user should be able to hand-edit models and create versions with the modified shapes.

Reply:
>_inputType.ValueCount is different than _numFeatures.

If `featureToInputMap` is null, I think they should be same, otherwise it's a size of `featureToInputMap`. Sorry, didn't properly check constructor of predictor.

>You mean the visualization command? With the new public interfaces, the user should be able to hand-edit models and create versions with the modified shapes.
I just find it strange what we have nested internal command in predictor. And only reason why it's nested is because we using some internal guts of predictor. I understand what you can do hand-edits, but can user visualize model same way how our visualize command work? If not, it's probably sign what we didn't expose enough.

Reply:
Ah, I see what you are saying. The visualization command is for an internal tool that may not be released.